### PR TITLE
chore: create proxy instance once per provider

### LINKED ISSUE #187: Create proxy per pass though route once
Currently new proxy for pass though routes is created for each request.
It should be created only once for each route.

### PR DESCRIPTION
**Pull request by pawbana**
Changes `newPassthroughRouter` so it created proxy per provider not on each request.
Replaced `Director` in `httputil.ReverseProxy` with `Rewrite` as Director is being [deprecated](https://cs.opensource.google/go/go/+/refs/tags/go1.26.2:src/net/http/httputil/reverseproxy.go;l=185) in next go version.

Fixes: https://github.com/coder/aibridge/issues/187

### CODE REVIEW

File: passthrough.go [hunk 1]
```diff
@@ -21,7 +20,39 @@ import (
 
 // newPassthroughRouter returns a simple reverse-proxy implementation which will be used when a route is not handled specifically
 // by a [intercept.Provider].
+// A single reverse proxy is created per provider and reused across all requests.
 func newPassthroughRouter(prov provider.Provider, logger slog.Logger, m *metrics.Metrics, tracer trace.Tracer) http.HandlerFunc {
+	provBaseURL, err := url.Parse(prov.BaseURL())
+	if err != nil {
+		return newInvalidBaseURLHandler(prov, logger, m, tracer, err)
+	}
+	if _, err := url.JoinPath(provBaseURL.Path, "/"); err != nil {
```
Comment (ssncferreira) [inline passthrough.go:30]: Why is this needed? If `provBaseURL` is properly set, this will never fail, right? 🤔

File: passthrough.go [hunk 2]
```diff
@@ -33,88 +64,48 @@ func newPassthroughRouter(prov provider.Provider, logger slog.Logger, m *metrics
...
-			},
-			ErrorHandler: func(rw http.ResponseWriter, req *http.Request, e error) {
-				logger.Warn(req.Context(), "reverse proxy error", slog.Error(e), slog.F("path", req.URL.Path))
-				http.Error(rw, "upstream proxy error", http.StatusBadGateway)
-			},
-		}
+	if prior, ok := pr.In.Header["X-Forwarded-For"]; ok {
+		pr.Out.Header["X-Forwarded-For"] = append([]string(nil), prior...)
+	}
+	pr.SetXForwarded()
```
Comment (ssncferreira) [inline passthrough.go:81]: Slightly confused with this, IIUC SetXForwarded sets all `X-Forwarded` headers right? `For`, `Host` and `Proto`, so why do we need the first if?

nit: might be worth adding a small comment

File: passthrough.go [hunk 3]
```diff
@@ -33,88 +64,48 @@ func newPassthroughRouter(prov provider.Provider, logger slog.Logger, m *metrics
...
+		pr.Out.Header.Set("User-Agent", "aibridge") // TODO: use build tag.
+	}
+
+	// Inject provider auth.
+	prov.InjectAuthHeader(&pr.Out.Header)
+}
+
+// newInvalidBaseURLHandler returns a handler that always returns 502 because
+// the provider's base URL is invalid.
+func newInvalidBaseURLHandler(prov provider.Provider, logger slog.Logger, m *metrics.Metrics, tracer trace.Tracer, baseURLErr error) http.HandlerFunc {
```
Comment (ssncferreira) [inline passthrough.go:97]: Nit: this duplicates the metrics/tracing from the happy-path handler (https://github.com/coder/aibridge/pull/267/changes#diff-b2b968f6cc08ffeda242a5870824976aa457930e6c2ef8b0ea80e747540dcd56R56). 
Maybe we could extract the common setup into a wrapper.

File: passthrough_test.go [hunk 4]
```diff
@@ -85,3 +130,161 @@ func TestPassthroughRoutes(t *testing.T) {
...
+				"X-Forwarded-For":   {"2.2.2.2, 3.3.3.3, 1.1.1.1"},
+				"User-Agent":        {"aibridge"},
+			},
+		},
+		{
+			name:          "omits_forwarded_for_when_remote_addr_is_not_parseable",
+			reqPath:       "http://client-host/chat",
+			reqRemoteAddr: "not-a-socket-address",
+			reqHeaders: http.Header{
+				"X-Forwarded-For": {"1.1.1.1"},
```
Comment (ssncferreira) [inline passthrough_test.go:233]: Why isn't this header in the `expectHeader` set?

File: passthrough_test.go [hunk 5]
```diff
@@ -85,3 +130,161 @@ func TestPassthroughRoutes(t *testing.T) {
...
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			r := httptest.NewRequest(http.MethodGet, tc.reqPath, nil)
+			maps.Copy(r.Header, tc.reqHeaders)
+			r.RemoteAddr = tc.reqRemoteAddr
+			if tc.reqTLS {
```
Comment (ssncferreira) [inline passthrough_test.go:252]: nit: I don't think we have a test where this is true 🤔

**Review by ssncferreira — COMMENTED** [review]
Overall LGTM, just some questions to make sure I understand this change 🙂

File: passthrough.go (hunk 1 above)
Comment (pawbana) [reply passthrough.go:30]: This one is interesting, `url.Parse` doesn't verify that path part of URL is properly encoded, that is what `url.JoinPath` is for.
It may be a bit pedantic but was already tested in `passthrough_route_break_parse_base_url` and I wanted to keep checks the same.

File: passthrough.go (hunk 2 above)
Comment (pawbana) [reply passthrough.go:81]: From [documentation](https://pkg.go.dev/net/http/httputil#ProxyRequest.SetXForwarded)

> To append to the inbound request's X-Forwarded-For header (the default behavior of [ReverseProxy](https://pkg.go.dev/net/http/httputil#ReverseProxy) when using a Director function), copy the header from the inbound request before calling SetXForwarded

I'll add a comment

File: passthrough_test.go (hunk 4 above)
Comment (pawbana) [reply passthrough_test.go:233]: As test case name suggest: `omits_forwarded_for_when_remote_addr_is_not_parseable`

`reqRemoteAddr` is set to not proper value. This is weird edge case that probably will never happen but I think it is better to point it out.

File: passthrough_test.go (hunk 5 above)
Comment (pawbana) [reply passthrough_test.go:252]: oh good catch, must have removed it by mistake. Will add back.

File: passthrough_test.go (hunk 4 above)
Comment (ssncferreira) [reply passthrough_test.go:233]: Right, but `1.1.1.1` is a valid address; it should still be kept in the header, no?

Comment (pawbana) [reply passthrough_test.go:233]: I'm not exactly sure why library decides to drop the header but maybe due to a security concerns?
If RemoteAddr / last hop is unknown it is better to drop whole header then forward it with missing hop.
Maybe I should remove this test as it tests library rather then our code.

Comment (ssncferreira) [reply passthrough_test.go:233]: Yeah, I think this is a very edge case and not sure it can even occur in production. I would suggest either removing the test, or if we decide to keep it add a comment explaining this for future reference

**Review by ssncferreira — APPROVED** [review]
LGTM 💪 thank you for addressing the comments!
